### DISCUSSION TITLE
Can I use external C libraries to run code on Crazyflie firmware?

### DISCUSSION THREAD

**Discussion by tjt22** [bitcraze/discussions#414] [Q&A]
Hi there!

I'm trying to read values from an Infineon TLI493D-W2BW hall effect sensor over I2C with the Crazyflie 2.1. I've been trying to use the commands from i2cdev.h to read values from the sensor with no success. The manufacturer of the sensor provided working code in C, but it requires their custom library. Am I able to import their library into the crazyflie firmware somehow and run their custom commands? Or do I have to use i2cdev.h for any interaction with i2c devices?

**Comment by ataffanel** [comment]
Pulling external library is possible as long as they are designed to run in an embedded system (which I would expect in that case). For example we are already pulling and compiling CMSIS math lib, FreeRTOS and libdw100 this way from git submodules: https://github.com/bitcraze/crazyflie-firmware/blob/master/vendor/Kbuild and other libs are copied in our git and compiled there: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/lib/Kbuild (including the ST vl53l1 driver which is also an I2C sensor driver).

So, if your driver comes as source code there should be no problem compiling it in both in the firmware repos or as an out of tree build, all you have to do is to add it to a KBuild and add the includes folders [to the Makefile](https://github.com/bitcraze/crazyflie-firmware/blob/5be4c083aa6700ee4e79286eb6d755e22520b5be/Makefile#L45).

Note that, in the case of the vl53l1 driver, we are still using i2cdev.h to implement the [communication functions of the driver](https://github.com/bitcraze/crazyflie-firmware/blob/5be4c083aa6700ee4e79286eb6d755e22520b5be/src/drivers/src/vl53l1x.c#L143).

**Comment by tjt22** [comment]
Ok thanks! The library comes in the form of a zip file. Would I just put this file in the same folder that Kbuild is in and then reference it in the file?
[Infineon-3D_Magnetic_Sensors_TLx493D_Library-Software-v01_03-EN.zip](https://github.com/bitcraze/discussions/files/10105893/Infineon-3D_Magnetic_Sensors_TLx493D_Library-Software-v01_03-EN.zip)


To make sure I have it right, can you verify that these are the steps I need to take?:
- Add a reference to the Kbuild in the master/src/lib path. Something like 'obj-y += Infineon-3D_Magnetic_Sensors_TLx493D_Library-Software-v01_03-EN.o'
- Add the includes: `-INCLUDES += -I$(srctree)/path of zip file'


I'm still a beginner at OS and C so I really appreciate the help!

**Reply by krichardsson** [reply]
You need to unzip the file so that you get `.c` and `.h` files first, this is what the compiler can use.

> Add a reference to the Kbuild in the master/src/lib path. Something like 'obj-y += Infineon-3D_Magnetic_Sensors_TLx493D_Library-Software-v01_03-EN.o'

You need to add one `obj-y += ...` statement for each `.c` file (that is used) in your library (`.c` files are compiled into `.o` files - object files). The make system is sort of doing things "backwards", these lines tells the make system that the object file is required and this will force it to compile the `.c` file to produce the missing `.o` file.

> Add the includes: `-INCLUDES += -I$(srctree)/path of zip file'

Your `-INCLUDES += ...` should point at the directory where the `.h` files are located. It is possible that you have multiple directories with header files and that you have to add multiple `-I$(LIB)/...` statements. This tells the compiler where to look for header file when there is a `#include` statement in a `.c` or `.h` file. Note that you should use `$(LIB)` if you place your files in `src/lib`, you can use `FatFS` for instance as a template.
